### PR TITLE
add `AddExchangeToBorder`, `CopyGuardToExchange`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- copy kernel and start routine for `insertField` and `bashField` from `FieldJ` to own functors
- remove `KernelBash*` and `KernelInsert*` from `FieldJ.kernel` and `FieldTmp.kernel`
- `FieldJ.tpp` and `FieldTmp.tpp` use new functors to copy and add exchange data

The algorithm of the kernel is not touched, only the usage of `simDim` is removed and `SuperCellSize` is taken from the type `T_Mapper`, also the naming for the boxes is changed.
This change avoid that we duplicate the code only for FieldJ and FieldTmp.

@Heikman, @n01r, @BenjaminW3 and @ax3l The kernel implementation is moved from PIConGPU `FieldJ.kernel` to PMacc, therefore the license needs to be changed. **Do you agree with the license change from GPLv3 only to GPLv3+LGPLv3?**

- [x] rebase against #2090 
- [x] KHI test (check HDF5 derived fields)

### CODE REVIEW

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            T_Extent const & exchangeSize,
+            T_Extent const & direction,
+            T_Mapping const & mapper
+        ) const
+        {
+            using namespace mappings::threads;
+
+            using SuperCellSize = typename T_Mapping::SuperCellSize;
+
+            /* number of cells in a superCell */
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: //

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            constexpr uint32_t numCells = PMacc::math::CT::volume< SuperCellSize >::type::value;
+            constexpr uint32_t numWorkers = T_numWorkers;
+            constexpr int dim = T_Mapping::Dim;
+
+            uint32_t const workerIdx = threadIdx.x;
+
+            DataSpace< dim > const  blockCell(
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: double space

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) )
+                    * SuperCellSize::toRT()
+            );
+
+            /*origin in area from local GPU*/
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: //

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                    numCells,
+                    numWorkers
+                >
+            >{ workerIdx }(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    /* cell index within the superCell */
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: //

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+
+namespace PMacc
+{
+namespace fields
+{
+namespace operations
+{
+
+    /** add the intermediate to the local box
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: Be less cryptic: *Add particles from a received temporary buffer (exchange) to the local particle box (border)*

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                        destBox( targetCell ) += exchangeBox( sourceCell );
+                }
+            );
+        }
+    };
+
+
+    /** add a exchange buffer to the border of the local buffer
+     *
+     * CopyGuardToExchange is the contrarious operation for the neighboring
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: contrarious -> opposite

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+     * device to create a exchange which can be added with this functor.
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: a*n*

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            using SuperCellSize = T_SuperCellSize;
+
+            constexpr int dim = T_SuperCellSize::dim;
+
+            using MappingDesc = MappingDescription<
+                dim,
+                SuperCellSize
+            >;
+
+            /* use only the x dimension to determine the number of supercells in the guard
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: maybe caps `GUARD`

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+             * PMacc restriction: all dimension must have the some number of guarding
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: if you don't put a `.` after the last sentence, add a newline or maybe even a `@warning`

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+
+namespace PMacc
+{
+namespace fields
+{
+namespace operations
+{
+
+    /** copy guarding cells to the intermediate box
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: - copy the +particles in+ guarding cells
- the -> an

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+     *
+     * @tparam T_numWorkers number of workers
+     */
+    template< uint32_t T_numWorkers >
+    struct KernelCopyGuardToExchange
+    {
+        /** copy guarding cells to the intermediate box
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: the -> an

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+            T_Extent const & exchangeSize,
+            T_Extent const & direction,
+            T_Mapping const & mapper
+        ) const
+        {
+            using namespace mappings::threads;
+
+            using SuperCellSize = typename T_Mapping::SuperCellSize;
+
+            /* number of cells in a superCell */
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: //

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+            constexpr int dim = T_Mapping::Dim;
+
+            uint32_t const workerIdx = threadIdx.x;
+
+            DataSpace< dim > const blockCell(
+                mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) ) *
+                SuperCellSize::toRT()
+            );
+
+            /*origin in area from local GPU*/
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: //

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+                    numCells,
+                    numWorkers
+                >
+            >{ workerIdx }(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    /* cell index within the superCell */
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: //

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+                    if( assignValue )
+                        exchangeBox( targetCell ) = srcBox( sourceCell );
+                }
+            );
+        }
+    };
+
+    /** copy guard of the local buffer to the exchange buffer
+     *
+     * AddExchangeToBorder is the contrarious operation for the neighboring
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: contrarious -> opposite

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    /* cell index within the superCell */
+                    DataSpace< dim > const cellIdx = DataSpaceOperations< dim >::template map< SuperCellSize >( linearIdx );
+                    DataSpace< dim > targetCell( blockCell + cellIdx );
+                    DataSpace< dim > sourceCell( targetCell - nullSourceCell );
+
+                    bool assignValue = true;
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: weird name: is this in the end an *assign* or an *add* (as the docs say above)

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+                    uint32_t const
+                )
+                {
+                    /* cell index within the superCell */
+                    DataSpace< dim > const cellIdx = DataSpaceOperations< dim >::template map< SuperCellSize >( linearIdx );
+
+                    DataSpace< T_Mapping::Dim > const sourceCell( blockCell + cellIdx );
+                    DataSpace< dim > targetCell( sourceCell - nullSourceCell );
+
+                    bool assignValue = true;
```
Comment (ax3l) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: weird name: is this in the end an *assign* or an *add* (as the docs say above)

File: src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp (hunk 16 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/fields/operations/AddExchangeToBorder.hpp]: It is a `add` (I will rename it)

File: src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp (hunk 17 above)
Comment (psychocoderHPC) [inline src/libPMacc/include/fields/operations/CopyGuardToExchange.hpp]: Assign is correct at this place, but I will rename it to `copy`

**Pull request merged**